contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -127,20 +131,46 @@ pub fn call_contract(
 
 pub fn route_messages(
     storage: &mut dyn Storage,
+    querier: QuerierWrapper,
     sender: Addr,
     msgs: Vec<Message>,
 ) -> Result<Response<nexus::execute::Message>> {
     let Config {
-        chain_name, router, ..
+        chain_name,
+        router,
+        nexus,
     } = state::load_config(storage);
+
     let router = Router::new(router);
+    let client: nexus::Client = client::CosmosClient::new(querier).into();
 
-    if sender == router.address {
-        Ok(prepare_msgs_for_execution(storage, chain_name, msgs)?)
-    } else {
-        // Messages initiated via call contract can be routed again
-        Ok(route_to_router(storage, &router, msgs)?)
-    }
+    msgs.into_iter()
+        .group_by(|msg| msg.destination_chain.to_owned())
+        .into_iter()
+        .try_fold(Response::new(), |mut acc, (dest_chain, msgs)| {
+            let response = match determine_routing_destination(&client, &dest_chain, &chain_name)? {
+                // Ensure only router routes to Axelarnet
+                RoutingDestination::Axelarnet => {
+                    ensure!(sender == router.address, Error::InvalidRoutingDestination);
+                    prepare_msgs_for_execution(storage, chain_name.clone(), msgs.collect())
+                }
+                // Ensure only router routes to Nexus
+                RoutingDestination::Nexus => {
+                    ensure!(sender == router.address, Error::InvalidRoutingDestination);
+                    route_messages_to_nexus(&client, &nexus, msgs.collect())
+                }
+                // Allow re-routing of CallContract initiated messages
+                RoutingDestination::Router => {
+                    ensure!(sender != router.address, Error::InvalidRoutingDestination);
+                    route_to_router(storage, &router, msgs.collect())
+                }
+            }?;
```

Comment:
I'd prefer for these functions `prepare_msgs_for_execution`, `route_messages_to_nexus`, and `route_to_router` to return a tuple of messages and events rather than a response.